### DISCUSSION TITLE
Synths: click when repeating same note

### DISCUSSION THREAD

**Discussion by nekotaro6** [ryukau/VSTPlugins#57] [General]
Hello. Here is something I noticed (not sure if it's a bug or not):
In some DAWs, polyphonic synths produce a click when the same MIDI note is repeated. Perhaps, there is something going on with voice being cut too abruptly.

This happens in Reaper, Ableton Live, Bitwig (although I checked in Live and Bitwig a while ago).
In FL Studio and Cubase there are no clicks.
Environment is Windows 11, 256 buffer, 48k sample rate.

Observed in: EnvelopedSine, IterativeSinCluster, CubicPadSynth, SyncSawSynth (perhaps some others too).

Attached a video demonstration in Reaper with EnvelopedSine and CubicPadSynth.
In EnvelopedSine it's easy to reproduce by choosing preset LongDecay and turning Phaser mix down.
Note, that if `Unison` is enabled in EnvelopedSine, click occurs only in left channel.
In CubicPadSynth turn Gain Envelope Attack up.

While this is not something critical, I wonder if a fix can be possible?

https://github.com/user-attachments/assets/250c928a-1028-4c42-8eb6-f83c54aae977

**Comment by ryukau** [comment]
Hi. Thanks for the report.

This is an intended behavior, but it's better to change. I'll work on it.

Following zip contains a possible fix. Please let me know if it removes click or not.

[EnvelopedSine.zip](https://github.com/user-attachments/files/17758127/EnvelopedSine.zip)

Edit: I'll work on other plugins after confirmation.

Edit2: The zip contains VST 3 for Windows.

**Comment by nekotaro6** [comment]
Thank you! The clicks are gone now.

**Comment by ryukau** [comment]
Thanks for confirmation. I'll work on the rest.

The root cause is an incomplete implementation of VST 3 note event on DAW side. Last time I checked, Live, REAPER, and Studio One were sending MIDI note number as VST 3 note ID. It seems that Bitwig is also doing this.

On my plugins, the intended behavior was to control the click with note ID. Typically, note ID changes for each note, but if a DAW is capable to set specific note ID, then we could have an additional expression.

The rest of this comment is documentation. You can ignore it.

<details>
I guess this is a good time to document the behavior of note ID and its purpose. In my opinion, VST 3 is doing the right thing here, and this is a shortcoming of MIDI specifications (both version 1 and 2, last time I checked).

Consider following sequence of note events. Time flows from top to bottom:

1. C4 note-on
2. C4 note-on
3. C4 note-off
4. C4 note-off

This is ambiguous because we can't know which note-on corresponds to which note-off. Possible combinations are:

- (1, 3), (2, 4)
- (1, 4), (2, 3)

With note ID, the sequence of note events changes to the following:

1. id: x, C4 note-on
2. id: y, C4 note-on
3. id: y, C4 note-off
4. id: x, C4 note-off

Now that we have ID for corresponding note-on and note-off, there's no ambiguity. Only possible combination is:

- (1, 4), (2, 3)

In MIDI, MIDI channel could be used as note ID. However, 16 channels aren't sufficient in my opinion.

In FL Studio, note ID is an incremental counter. The ID goes up by 1 for each new pair of note-on and note-off.
</details>

**Comment by ryukau** [comment]
The fix was to change the following line:

https://github.com/ryukau/VSTPlugins/blob/5084d97a31704a9ff3a56a0cada50e7fa74f5ba9/EnvelopedSine/source/dsp/dspcore.cpp#L332

to:

```c++
if (notes[ndx].state == NoteState::rest) { 
```

**Comment by ryukau** [comment]
The fix is applied to other polyphonic synthesizers. @nekotaro6, would you be able to confirm the fix? Below is a download link.

[polysynth_fix_20241116.zip](https://github.com/user-attachments/files/17780462/polysynth_fix_20241116.zip)

Gain envelopes of SyncSawSynth and IterativeSinCluster are changed to reduce the click.

Following synths can be ignored.

- ClangSynth
- CollidingCombSynth
- EnvelopedSine (tested)

Click is hard to notice on ClangSynth and CollidingCombSynth because the texture is added by delays.

<details>
<summary>Testing of delay based synths</summary>

### ClangSynth
1. Turn off `FDN`.
2. Raise `Gain -> Release` on top left.
3. Raise `Oscillator -> Denom. Slope` on top middle.
4. Press `Refresh Waveform`.

### CollidingCombSynth
It's hard to distinguish the click.

0. Insert limiter after CollidingCombSynth.
1. Turn off `Compressor`.
2. Lower `String -> LP Cutoff` on middle.
3. Raise `Unison -> nUnison` to 8 (maximum).

</details>

**Comment by nekotaro6** [comment]
I can confirm, that on all synths provided, the click issue is fixed. Thank you for your work!

**Comment by ryukau** [comment]
@nekotaro6 Thank you. The changes are released as version 0.65.0.

Edit: I'll leave open this discussion. If anything is left, feel free to add a comment.
